File: tyhp-lang/phpstorm/src/main/kotlin/com/tyhp/lang/settings/SettingsCore.kt
package com.tyhp.lang.settings

/**
 * Pure helpers for `tyhp.*` settings. Keep IntelliJ APIs out of this file so unit
 * tests can run as plain JVM tests without `runIde`.
 *
 * Key names match the VS Code contribution points (`tyhp.path`, `tyhp.binary.*`, …).
 */

enum class InstallMode(val value: String) {
    PATH("path"),
    GLOBAL("global"),
    EXTENSION("extension");

    companion object {
        fun parse(value: String?): InstallMode =
            entries.firstOrNull { it.value == value } ?: PATH
    }
}

/** Where to persist `tyhp.path` after a PATH probe or install. */
enum class PathWriteTarget {
    Application,
    Project,
}

/**
 * Snapshot of `tyhp.path` at each scope. A non-null [projectValue] means the
 * project has an override (even if the string is empty), matching VS Code’s
 * `inspect().workspaceValue !== undefined`.
 */
data class InspectedPath(
    val applicationValue: String? = null,
    val projectValue: String? = null,
)

fun isPathUnset(value: String?): Boolean = value == null || value.trim().isEmpty()

fun parseInstallMode(value: String?): InstallMode = InstallMode.parse(value)

/**
 * Effective `tyhp.path`: a project override wins even when it is the empty
 * string (matching VS Code `inspect().workspaceValue !== undefined`).
 */
fun effectiveTyhpPath(inspect: InspectedPath): String {
    val project = inspect.projectValue
    if (project != null) {
        return project.trim()
    }
    return (inspect.applicationValue ?: "").trim()
}

/**
 * User (application) by default unless a project override already exists.
 */
fun pathWriteTarget(inspect: InspectedPath?): PathWriteTarget {
    if (inspect?.projectValue != null) {
        return PathWriteTarget.Project
    }
    return PathWriteTarget.Application
}

fun normalizePinnedVersion(value: String?): String = (value ?: "").trim()

/** GitHub tags are `vX.Y.Z…`; settings may omit the `v`. */
fun normalizeReleaseTag(value: String): String {
    val trimmed = value.trim()
    if (trimmed.isEmpty()) {
        return ""
    }
    return if (trimmed.startsWith("v") || trimmed.startsWith("V")) trimmed else "v$trimmed"
}

fun tagsMatch(a: String, b: String): Boolean {
    val left = normalizeReleaseTag(a)
    val right = normalizeReleaseTag(b)
    if (left.isEmpty() || right.isEmpty()) {
        return false
    }
    return left.equals(right, ignoreCase = true)
}

File: tyhp-lang/vscode/src/config/settingsCore.ts
/**
 * Pure helpers for `tyhp.*` settings. Keep vscode imports out of this file
 * so unit tests can run under `node --test` without the extension host.
 */

export type InstallMode = "path" | "global" | "extension";

export const INSTALL_MODES: readonly InstallMode[] = ["path", "global", "extension"];

/** Mirrors `vscode.ConfigurationTarget` so tests do not import `vscode`. */
export enum SettingsScope {
    Global = 1,
    Workspace = 2,
    WorkspaceFolder = 3,
}

export interface InspectedSetting<T> {
    globalValue?: T;
    workspaceValue?: T;
    workspaceFolderValue?: T;
}

export function isPathUnset(value: string | undefined | null): boolean {
    return value === undefined || value === null || value.trim() === "";
}

export function parseInstallMode(value: unknown): InstallMode {
    if (value === "global" || value === "extension" || value === "path") {
        return value;
    }
    return "path";
}

/**
 * Where to persist `tyhp.path` after a PATH probe or install.
 * User (Global) by default unless a workspace / folder override already exists.
 */
export function pathWriteTarget(inspect: InspectedSetting<string> | undefined): SettingsScope {
    if (!inspect) {
        return SettingsScope.Global;
    }
    if (inspect.workspaceFolderValue !== undefined) {
        return SettingsScope.WorkspaceFolder;
    }
    if (inspect.workspaceValue !== undefined) {
        return SettingsScope.Workspace;
    }
    return SettingsScope.Global;
}

export function normalizePinnedVersion(value: string | undefined | null): string {
    return (value ?? "").trim();
}

/** GitHub tags are `vX.Y.Z…`; settings may omit the `v`. */
export function normalizeReleaseTag(value: string): string {
    const trimmed = value.trim();
    if (trimmed === "") {
        return "";
    }
    return trimmed.startsWith("v") || trimmed.startsWith("V") ? trimmed : `v${trimmed}`;
}

export function tagsMatch(a: string, b: string): boolean {
    const left = normalizeReleaseTag(a);
    const right = normalizeReleaseTag(b);
    if (left === "" || right === "") {
        return false;
    }
    return left.toLowerCase() === right.toLowerCase();
}

/**
 * Value the user actually stored (folder → workspace → user). Package.json
 * defaults are not stored, so this is `undefined` when the setting is unset.
 */
export function explicitInspectedValue<T>(inspect: InspectedSetting<T> | undefined): T | undefined {
    if (!inspect) {
        return undefined;
    }
    if (inspect.workspaceFolderValue !== undefined) {
        return inspect.workspaceFolderValue;
    }
    if (inspect.workspaceValue !== undefined) {
        return inspect.workspaceValue;
    }
    if (inspect.globalValue !== undefined) {
        return inspect.globalValue;
    }
    return undefined;
}
